src/stateManager/DisputeValidationService.ts

```diff
@@ -277,28 +296,161 @@ export default class DisputeValidationService {
+            }
+            // [check] isTimedoutTooEarly
+            const timeoutTimestamp = Number(
+                await this.diamondStateMachine.localDiamondContract.getDisputeWindowCreationTimestamp(
+                    dispute.input.channelId,
+                    dispute.input.forkId
+                )
+            );
+            if (!timeoutTimestamp)
+                throw new Error(
+                    "Timeout timestamp not found, dispute state not synced locally"
+                );
+            // TODO - this doesn't account for race condition (us not aware of on-chain calldata)
+            const previousBlockOrSnapshot =
+                this.storage.getPreviousBlockOrSnapshot(cooridnates);
+            let previousTimestamp = 0;
+            if (previousBlockOrSnapshot.stateSnapshot) {
+                previousTimestamp =
+                    previousBlockOrSnapshot.stateSnapshot.timestamp;
+            } else {
+                // previous block
+                const previousBlock = previousBlockOrSnapshot.block!;
+                const onChainSignature = dispute.input.timeout
+                    .participantSignatureOnPreviousBlock as Signature;
+                if (!onChainSignature || onChainSignature === "0x") {
+                    // siganture not posted on-chain, so time is not forfeited
+                    previousTimestamp = previousBlock.currentTimestamp;
+                } else {
+                    // signature exists on-chain, verify it's from the timedout participant
+                    const retrievedAddress = previousBlock.signatureToAddress(
+                        dispute.input.timeout
+                            .participantSignatureOnPreviousBlock as Signature
+                    );
+                    if (retrievedAddress == dispute.input.timeout.participant) {
+                        // signature is valid, so extra time is forfeited
+                        previousTimestamp = previousBlock.currentTimestamp;
+                    } else {
+                        // signature is invalid, so extra time is not forfeited
+                        previousTimestamp = previousBlock.getRelevantTimestamp(
+                            dispute.input.timeout.participant
+                        );
+                    }
+                }
+            }
+            // previousTimestamp is now correctly set
+            // TODO - think if it's <= or <
+            if (
+                timeoutTimestamp <
+                previousTimestamp +
+                    this.stateManager.getTimeoutWaitTimeSeconds()
+            ) {
+                this.disputeFraudProofService.createTimeoutTooEarly(
+                    dispute,
+                    disputeAuditingData,
+                    previousBlockOrSnapshot?.block?.onChainTimestamp
+                );
+                return false;
+            }
+
+            // [check] N/N Threshold
+            if (block && block.didEveryoneSign(participants)) {
```

Comment:
better handle the case of block undefined at the top, right after `storage.blocks.getBlock`